CMR-8679: Enable POST requests on the /access-control/permissions endpoint

dpesall opened the pull request:
In the order broker, we will be making use of the /access-control/permissions endpoint. Currently the endpoint only accepts GET requests, which is limiting when requesting permissions for large sets of concept ids. Enabling POST requests would remove this limitation.

daniel-zamora commented on access-control-app/src/cmr/access_control/api/routes.clj:
```diff
@@ -368,7 +368,11 @@
-             (get-permissions ctx params)))
+             (get-permissions ctx params))
+
+        (POST "/"
```
is the idea that they will be submitting a post request body? or are they wanting to hit a POST endpoint while using query params?

dpesall replied:
Yes, you're right. I misunderstood the requirement. I've made the change to use a POST request body for providing the parameters.

ygliuvt reviewed (changes requested):
The ticket is asking for POST with request body. This is similar to [Search with POST](https://cmr.earthdata.nasa.gov/search/site/docs/search/api.html#search-with-post) in CMR search.

dpesall commented:
> The ticket is asking for POST with request body. This is similar to [Search with POST](https://cmr.earthdata.nasa.gov/search/site/docs/search/api.html#search-with-post) in CMR search.

Fixed.

daniel-zamora commented:
I think we need some test cases, updated docs and it might not be entirely necessary to split the call to get-permissions out into the get-permissions-with-post function.

eudoroolivares2016 commented on access-control-app/docs/api.md:
````diff
@@ -863,6 +864,19 @@ Content-Type: application/json;charset=ISO-8859-1
+
+```
+curl -i -XPOST -H "Content-Type: application/json" "%CMR-ENDPOINT%/permissions" -d '{ "user_type" : "guest", "concept_id" : ["C1200000011-PROV1", "C1200000031-PROV1"] }'
+
````
```suggestion

__Sample response__
```

dpesall commented:
> I think we need some test cases, updated docs and it might not be entirely necessary to split the call to get-permissions out into the get-permissions-with-post function.

Updated API docs, consolidated permissions POST and GET function calls, and added tests.

daniel-zamora commented on access-control-app/int-test/cmr/access_control/int_test/permission_check_test.clj:
```diff
@@ -920,6 +921,59 @@
+  "Helper to get permissions with the current context and the specified username string or user type keyword and concept ids."
+  [user & concept-ids]
+  (json/parse-string
+    (ac/get-permissions
```
Non macro indentation is 1

dpesall replied:
Oops, that function ended up being not needed since I accomplished what I needed without changing that function. Deleted now.

daniel-zamora commented on access-control-app/int-test/cmr/access_control/int_test/permission_check_test.clj:
```diff
@@ -920,6 +921,59 @@
             "user1" ["update" "delete"]
             "user2" []))))))
 
+(defn get-permissions-2
```
nit, get-permissions-2 needs to be descriptive.  Maybe get-user-concept-id-permissions or something like that.  Mostly just remove the -2

daniel-zamora commented on access-control-app/int-test/cmr/access_control/int_test/permission_check_test.clj:
```diff
@@ -920,6 +921,48 @@
+                         "] }")]
+
+    (testing "permissions endpoint allows get request"
+      (let [get-response (shell/sh "curl" "-g" "-i" get-url)
```
There should be test helpers that use a http client that you can use to make calls, instead of curls.

daniel-zamora replied:
YOu also do not need a GET test for this ticket, only POST.  You could piggy back on another test that already does a get, and simply repeat the test again for POST

dpesall replied:
Removed the GET test and modified the POST to use the http client.

jaybarra commented on system-int-test/test/cmr/system_int_test/access_control/permissions_route_test.clj:
```diff
@@ -0,0 +1,40 @@
+        coll3 (save-basic-collection "coll3")
+        coll4 (save-basic-collection "coll4")
+        permissions-url (url/access-control-permissions-url)
+        post-data-body (str "{ \"user_type\" : \"guest\", \"concept_id\" : ["
+                         "\"" coll1 "\", "
+                         "\"" coll2 "\", "
+                         "\"" coll3 "\", "
+                         "\"" coll4 "\""
+                         "] }")]
```
Would prefer if this was marshaled into JSON with cheshire.

dpesall replied:
Fixed

jaybarra reviewed (changes requested):
Can we add a test for an invalid query body to make sure we validate input as well?

ygliuvt reviewed (changes requested):
This ticket is asking for searching with "query parameters encoded form in POST request body". See the link I provided in a previous comment. This PR didn't provide this feature.

jceaser commented on access-control-app/src/cmr/access_control/api/routes.clj:
```diff
@@ -2,6 +2,7 @@
   "Defines the HTTP URL routes for the access-control API."
   (:require
    [cheshire.core :as json]
+   [clojure.walk :as w]
```
w is to short I thought, what about "walk"

dpesall replied:
Good point! Changed to `walk`

jceaser commented on system-int-test/test/cmr/system_int_test/access_control/permissions_route_test.clj:
```diff
@@ -0,0 +1,44 @@
+   [clj-http.client :as client]
+   [clojure.string :as string]
+   [clojure.test :refer :all]
+   [cmr.access-control.test.util :as u]
```
u is to short I thought

dpesall replied:
Changed to `ac-util`

jceaser commented on access-control-app/docs/api.md:
````diff
@@ -863,6 +864,19 @@ Content-Type: application/json;charset=ISO-8859-1
+Example POST request:
+
+```
+curl -i -XPOST -H "Content-Type: application/json" "%CMR-ENDPOINT%/permissions" -d 'user_type=guest&concept_id=C1200000003-PROV1&concept_id=C1200000004-PROV1'
````
can we add some new lines ("\") to break this up and keep it from needing a scroll?

dpesall replied:
Moved the query body to its own line.

ygliuvt commented on access-control-app/docs/api.md:
````diff
@@ -863,6 +864,20 @@ Content-Type: application/json;charset=ISO-8859-1
 {"C1200000000-PROV1": ["read"], "C1200000001-PROV1": []}
 ```
 
+When requesting permissions for large sets of concept ids, you can make a POST request to `%CMR-ENDPOINT%/permissions` and pass the query parameters in the POST body.
````
change `query parameters` to `query parameters in URL encoded query string`.

dpesall replied:
Fixed

ygliuvt commented on access-control-app/src/cmr/access_control/api/routes.clj:
```diff
@@ -368,7 +370,11 @@
+
+        (POST "/"
+              {ctx :request-context headers :headers body :body}
+              (get-permissions ctx (walk/keywordize-keys (codec/form-decode (slurp body))))))
```
No need to parse the query string yourself. Ring middleware has already put the query parameters in the `:params`.

ygliuvt commented on system-int-test/test/cmr/system_int_test/access_control/permissions_route_test.clj:
```diff
@@ -0,0 +1,44 @@
+(ns cmr.system-int-test.access-control.permissions-route-test
```
access-control integration tests are not in `system-int-test`, they are within the access-control project. See `cmr.access-control.int-test.permission-check-test` for existing `get-permissions` tests.

dpesall replied:
Moved

ygliuvt commented on system-int-test/src/cmr/system_int_test/utils/url_helper.clj:
```diff
@@ -691,6 +691,11 @@
   []
   (format "http://localhost:%s/acls" (transmit-config/access-control-port)))
 
+(defn access-control-permissions-url
```
This can be removed as we will add the tests in access-control app.

dpesall replied:
Removed

ygliuvt reviewed (changes requested):
Need to address the review comments.

dpesall commented on access-control-app/docs/api.md:
````diff
@@ -863,6 +864,20 @@ Content-Type: application/json;charset=ISO-8859-1
+Example POST request:
+
+```
+curl -i -XPOST -H "Content-Type: application/json" "http://localhost:3011/permissions" -d \
````
Fixed

ygliuvt commented on access-control-app/int-test/cmr/access_control/int_test/permission_check_test.clj:
```diff
@@ -23,6 +24,36 @@
+                        :content-type "application/x-www-form-urlencoded"})
+            body (get response :body)]
+
+        (is (string/includes? body coll1))
```
These checks are too opaque. We should either add a `get-permission-via-post` call in all the existing tests and verify they all return the same result as `get` call or in this test makes a get permissions call via `get` and verify the responses from `post` and `get` match at the least.

ygliuvt replied:
or remove this test and add a test case in at least one of the tests below to verify that `get` and `post` with the same query params return the same result.

dpesall replied:
I played with each option and decided to go with adding the `get` response to this test.

ygliuvt commented on access-control-app/src/cmr/access_control/api/routes.clj:
```diff
@@ -2,6 +2,7 @@
   "Defines the HTTP URL routes for the access-control API."
   (:require
    [cheshire.core :as json]
+   [clojure.walk :as walk]
```
Remove this line.

dpesall replied:
Removed

ygliuvt commented on access-control-app/src/cmr/access_control/api/routes.clj:
```diff
@@ -27,7 +28,8 @@
    [ring.middleware.nested-params :as nested-params]
-   [ring.middleware.params :as params])
+   [ring.middleware.params :as params]
+   [ring.util.codec :as codec])
```
Remove this line.

dpesall replied:
Removed
